services/horizon/internal/ingest/processors/contract_data.go

```diff
@@ -139,6 +150,106 @@ func AssetFromContractData(ledgerEntry xdr.LedgerEntry, passphrase string) *xdr.
 	return &asset
 }
 
+// ContractBalanceFromContractData takes a ledger entry and verifies if the ledger entry corresponds
+// to the balance entry written to contract storage by the Stellar Asset Contract. See:
+// https://github.com/stellar/rs-soroban-env/blob/5695440da452837555d8f7f259cc33341fdf07b0/soroban-env-host/src/native_contract/token/storage_types.rs#L11-L24
+func ContractBalanceFromContractData(ledgerEntry xdr.LedgerEntry, passphrase string) ([32]byte, *big.Int, bool) {
+	if ledgerEntry.Data.Type != xdr.LedgerEntryTypeContractData {
+		return [32]byte{}, nil, false
+	}
+
+	contractData := ledgerEntry.Data.MustContractData()
+	// we don't support asset stats for lumens
+	nativeAssetContractID, err := xdr.MustNewNativeAsset().ContractID(passphrase)
+	if err != nil || contractData.ContractId == nativeAssetContractID {
+		return [32]byte{}, nil, false
+	}
+
+	if contractData.Key.Type != xdr.ScValTypeScvObject {
+		return [32]byte{}, nil, false
+	}
+	keyObj := contractData.Key.MustObj()
+	if keyObj == nil {
+		return [32]byte{}, nil, false
+	}
+	if keyObj.Type != xdr.ScObjectTypeScoVec {
+		return [32]byte{}, nil, false
+	}
+	keyEnumVec := keyObj.MustVec()
+	if len(keyEnumVec) != 2 || !keyEnumVec[0].Equals(xdr.ScVal{Type: xdr.ScValTypeScvSymbol, Sym: &balanceMetadataSym}) {
+		return [32]byte{}, nil, false
+	}
+	if keyEnumVec[1].Type != xdr.ScValTypeScvObject {
+		return [32]byte{}, nil, false
+	}
+	addressObj := keyEnumVec[1].MustObj()
+	if addressObj == nil {
+		return [32]byte{}, nil, false
+	}
+	if addressObj.Type != xdr.ScObjectTypeScoAddress {
+		return [32]byte{}, nil, false
+	}
+	scAddress := addressObj.MustAddress()
+	if scAddress.Type != xdr.ScAddressTypeScAddressTypeContract {
+		return [32]byte{}, nil, false
+	}
+	holder := scAddress.MustContractId()
```

Comment:
I've found this parsing pattern to be a little easier to read:

```suggestion
	addressObj, ok := keyEnumVec[1].GetObj()
	if !ok || addressObj == nil {
		return [32]byte{}, nil, false
	}
	address, ok := addressObj.GetAddress()
	if !ok || address == nil {
		return [32]byte{}, nil, false
	}
	holder, ok := address.GetContractId()
	if !ok {
		return [32]byte{}, nil, false
	}
```

Basically, let `Get<Type>()` do the extra type-checking work for you! You don't have to change it here as its a buncha work, but for the future I find this cleaner.